tuning/trainercontroller/operations/logcontrol.py

```diff
@@ -0,0 +1,61 @@
+# Third Party
+from transformers import TrainingArguments
+from transformers.utils import logging
+
+# Local
+from .operation import Operation
+
+logger = logging.get_logger(__name__)
+logger.setLevel(level=logging.DEBUG)
+
+VALID_LOG_LEVELS = {
+    "ERROR": logging.ERROR,
+    "WARNING": logging.WARNING,
+    "INFO": logging.INFO,
+    "DEBUG": logging.DEBUG,
+}
+
+
+class LogControl(Operation):
+    """Operation that can be used to log useful information on specific events."""
+
+    def __init__(self, log_format: str, log_level: str, **kwargs):
+        """Initializes the HuggingFace controls. In this init, the fields with `should_` of the
+        transformers.TrainerControl data class are extracted, and for each of those fields, the
+        control_action() method's pointer is set, and injected as a class member function.
+
+        Args:
+            kwargs: List of arguments (key, value)-pairs
+        """
+        if log_level not in VALID_LOG_LEVELS:
+            raise ValueError(
+                "Specified log_level [%s] is invalid for LogControl" % (log_level)
+            )
+        self.log_level = VALID_LOG_LEVELS[log_level]
+        self.log_format = log_format
+        super().__init__(**kwargs)
+
+    def should_log(
+        self,
+        event_name: str,
+        tc_metrics: dict,
+        args: TrainingArguments,
+        **kwargs,
+    ):
+        """This method peeks into the stack-frame of the caller to get the action the triggered
+        a call to it. Using the name of the action, the value of the control is set.
+
+        Args:
+            control: TrainerControl. Data class for controls.
+            kwargs: List of arguments (key, value)-pairs
+        """
+        log_msg = self.log_format.format(
+            event_name=event_name,
+            tc_metrics=tc_metrics,
+            args=args,
+            **kwargs,
```

Comment:
should we flatten this to make rule and format to look similar?

Reply:
Could you please elaborate on this?

Reply:
@ashokponkumar I have addressed the above comment. Following are the keywords we need to avoid using as trainer controller metric names as they conflict when passed with `**kwargs`:
```
model 
tokenizer
optimizer
lr_scheduler
train_dataloader
eval_dataloader 
metrics
logs
```
Source is [HF Callback class definition](https://github.com/huggingface/transformers/blob/868d36d29ec132deeaaf8571b25b6a1b911d0145/src/transformers/trainer_callback.py#L261).